### <span style="color:red">**Task 1: Data Analysis & Understanding**</span>

In [1]:
import pandas as pd 
import pandasql as ps

In [14]:
customer_dataset = pd.read_excel('revenue_dataset.xlsx', sheet_name='customers')
invoice_dataset = pd.read_excel('revenue_dataset.xlsx', sheet_name='invoices')
position_dataset = pd.read_excel('revenue_dataset.xlsx', sheet_name='positions')

In [16]:
invoice_dataset['Zahlungsdatum'] = pd.to_datetime(invoice_dataset['Zahlungsdatum'], format='%Y%m-%d', errors='coerce')
invoice_dataset

ReNummer  SummeNetto  MwStSatz  ZahlungsbetragBrutto   KdNr  \
0       103601     11786.7         0                  0.00  78911   
1       103600       379.0         7                  0.00  79666   
2       103599       160.0         7                  0.00  30145   
3       103598       225.0         7                  0.00  10383   
4       103597      1260.0         7                  0.00  20213   
...        ...         ...       ...                   ...    ...   
1995    101606     37550.0         0              37842.17  78962   
1996    101605      3450.0         7               3691.50  20020   
1997    101604        94.5         7                101.12  20843   
1998    101603       375.0         7                401.25  20843   
1999    101602       145.0         0                145.00  81044   

      Summenebenkosten    ReDatum Zahlungsdatum  
0                 0.00 2025-04-04           NaT  
1                 0.00 2025-04-03           NaT  
2                 0.00 2025-04-04           NaT  
3                 0.00 2025-04-01           NaT  
4                 0.00 2025-04-03           NaT  
...                ...        ...           ...  
1995           -292.17 2024-11-04    2024-11-07  
1996              0.00 2024-11-04    2024-11-12  
1997              0.00 2024-11-04    2024-11-18  
1998              0.00 2024-11-04    2024-11-28  
1999              0.00 2024-11-04    2024-11-08  

[2000 rows x 8 columns]

In [17]:
invoice_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   ReNummer              2000 non-null   int64         
 1   SummeNetto            2000 non-null   float64       
 2   MwStSatz              2000 non-null   int64         
 3   ZahlungsbetragBrutto  1999 non-null   float64       
 4   KdNr                  2000 non-null   int64         
 5   Summenebenkosten      1998 non-null   float64       
 6   ReDatum               2000 non-null   datetime64[ns]
 7   Zahlungsdatum         1597 non-null   datetime64[ns]
dtypes: datetime64[ns](2), float64(3), int64(3)
memory usage: 125.1 KB


In [4]:
customer_dataset

id   Kdnr                                      Verlagsname     Region
0        5  20172                                   1. FC Nürnberg   Nürnberg
1       19  20137  Allgäuer Zeitung / Allgäuer Zeitungsverlag GmbH   Bodensee
2       27  20115                            Augsburger Allgemeine    München
3       69  10113                         Brainpool TV Productions  Rheinland
4       72  10154                 RFW / Redaktionsbüro Wipperfürth    Hamburg
..     ...    ...                                              ...        ...
596  14014  81391                                              PPF        NaN
597  14021  81398          VLAAMSE RADIO- EN TELEVISIEOMROEP (VRT)        NaN
598  14025  81402                           Umweltinstitut München        NaN
599  14031  81408                       Everprod - Groupe Elephant        NaN
600  14034  81411                                  Lennart Homeyer        NaN

[601 rows x 4 columns]

In [9]:
customer_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 601 entries, 0 to 600
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   id           601 non-null    int64 
 1   Kdnr         601 non-null    int64 
 2   Verlagsname  601 non-null    object
 3   Region       280 non-null    object
dtypes: int64(2), object(2)
memory usage: 18.9+ KB


In [6]:
position_dataset

id    ReId     KdNr  Nettobetrag   Bildnummer   VerDatum
0       4154495  102038  50292.0        19.00  168161043.0 2022-09-29
1       4154496  102038  50292.0        19.00  108514408.0 2022-09-12
2       4154497  102038  50292.0        19.00   55374589.0 2022-09-08
3       4154498  102038  50292.0        19.00   55374589.0 2022-09-08
4       4154499  102038  50292.0        19.00   57865918.0 2022-09-08
...         ...     ...      ...          ...          ...        ...
129087  5726949  103209  10813.0        12.00  100000000.0 2025-01-27
129088  5726950  103209  10813.0        12.00  100000000.0 2025-02-25
129089  5726951  103209  10813.0        12.00  100000000.0 2025-02-25
129090  5726952  103209  10813.0        12.00  100000000.0 2025-02-23
129091  5726953  103209  10813.0         0.24  100000000.0 2025-02-15

[129092 rows x 6 columns]

In [7]:
position_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 129092 entries, 0 to 129091
Data columns (total 6 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   id           129092 non-null  int64         
 1   ReId         129092 non-null  int64         
 2   KdNr         129091 non-null  float64       
 3   Nettobetrag  129091 non-null  float64       
 4   Bildnummer   129091 non-null  float64       
 5   VerDatum     129088 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(3), int64(2)
memory usage: 5.9 MB


### <span style="color:#ADD8E6">**Data Cleaning**</span>

#### <span style="color:orange">**Invoice Dataset**</span>

#Do data cleaning of each task and explain why you did it

invoice_null = invoice_dataset[invoice_dataset.isnull().sum() > 0]
invoice_null

In [21]:
invoice_null = invoice_dataset.isnull().sum()
invoice_null

ReNummer                  0
SummeNetto                0
MwStSatz                  0
ZahlungsbetragBrutto      1
KdNr                      0
Summenebenkosten          2
ReDatum                   0
Zahlungsdatum           403
dtype: int64

In [24]:
missing_payment_rows = invoice_dataset[
    invoice_dataset['ZahlungsbetragBrutto'].isna() | 
    invoice_dataset['Summenebenkosten'].isna()
]
missing_payment_rows

ReNummer  SummeNetto  MwStSatz  ZahlungsbetragBrutto   KdNr  \
19      103582       350.0         7                   0.0  10383   
21      103580     11000.0         7                   0.0  20608   
1270    102331      1297.5         7                   NaN  10957   

      Summenebenkosten    ReDatum Zahlungsdatum  
19                 NaN 2025-04-03           NaT  
21                 NaN 2025-04-03           NaT  
1270               0.0 2025-04-01           NaT

In [30]:
unpaid_invoices = invoice_dataset[invoice_dataset['Zahlungsdatum'].isna()]
unpaid_brutto_sum = unpaid_invoices['ZahlungsbetragBrutto'].sum()
print(f"The total unpaid brutto sum is: {unpaid_brutto_sum}")

The total unpaid brutto sum is: 11636.25


<span style="color: pink"> Our analysis revealed three critical data quality issues in the invoices dataset: </span>
- <span style="color: pink"> 1 missing gross payment amount (ZahlungsbetragBrutto)  </span>
-  <span style="color: pink">2 missing additional cost values (Summenebenkosten) </span>
-  <span style="color: pink">403 missing payment dates (Zahlungsdatum). </span>

<span style="color: pink"> These gaps impact financial reporting accuracy. The gross payment amounts and additional costs are derivable from existing fields, while payment dates represent factual business events that shouldn't be imputed.</span>


In [33]:
# Assume missing additional costs = 0
invoice_dataset['Summenebenkosten'] = invoice_dataset['Summenebenkosten'].fillna(0)

# ZahlungsbetragBrutto = (SummeNetto * (100 + MwStSatz) / 100) -> This is the amnount they have to pay after tax
invoice_dataset['ZahlungsbetragBrutto'] = invoice_dataset.apply(
    lambda row: row['SummeNetto'] * (1 + row['MwStSatz']/100)  # Corrected formula
    if pd.isna(row['ZahlungsbetragBrutto']) or row['ZahlungsbetragBrutto'] == 0
    else row['ZahlungsbetragBrutto'],
    axis=1
)

invoice_dataset


ReNummer  SummeNetto  MwStSatz  ZahlungsbetragBrutto   KdNr  \
0       103601     11786.7         0              11786.70  78911   
1       103600       379.0         7                405.53  79666   
2       103599       160.0         7                171.20  30145   
3       103598       225.0         7                240.75  10383   
4       103597      1260.0         7               1348.20  20213   
...        ...         ...       ...                   ...    ...   
1995    101606     37550.0         0              37842.17  78962   
1996    101605      3450.0         7               3691.50  20020   
1997    101604        94.5         7                101.12  20843   
1998    101603       375.0         7                401.25  20843   
1999    101602       145.0         0                145.00  81044   

      Summenebenkosten    ReDatum Zahlungsdatum  
0                 0.00 2025-04-04           NaT  
1                 0.00 2025-04-03           NaT  
2                 0.00 2025-04-04           NaT  
3                 0.00 2025-04-01           NaT  
4                 0.00 2025-04-03           NaT  
...                ...        ...           ...  
1995           -292.17 2024-11-04    2024-11-07  
1996              0.00 2024-11-04    2024-11-12  
1997              0.00 2024-11-04    2024-11-18  
1998              0.00 2024-11-04    2024-11-28  
1999              0.00 2024-11-04    2024-11-08  

[2000 rows x 8 columns]

In [39]:
invoice_dataset['PaymentStatus'] = invoice_dataset['Zahlungsdatum'].apply(lambda x: 'Paid' if pd.notnull(x) else 'Unpaid')
invoice_dataset

ReNummer  SummeNetto  MwStSatz  ZahlungsbetragBrutto   KdNr  \
0       103601     11786.7         0              11786.70  78911   
1       103600       379.0         7                405.53  79666   
2       103599       160.0         7                171.20  30145   
3       103598       225.0         7                240.75  10383   
4       103597      1260.0         7               1348.20  20213   
...        ...         ...       ...                   ...    ...   
1995    101606     37550.0         0              37842.17  78962   
1996    101605      3450.0         7               3691.50  20020   
1997    101604        94.5         7                101.12  20843   
1998    101603       375.0         7                401.25  20843   
1999    101602       145.0         0                145.00  81044   

      Summenebenkosten    ReDatum Zahlungsdatum PaymentStatus  
0                 0.00 2025-04-04           NaT        Unpaid  
1                 0.00 2025-04-03           NaT        Unpaid  
2                 0.00 2025-04-04           NaT        Unpaid  
3                 0.00 2025-04-01           NaT        Unpaid  
4                 0.00 2025-04-03           NaT        Unpaid  
...                ...        ...           ...           ...  
1995           -292.17 2024-11-04    2024-11-07          Paid  
1996              0.00 2024-11-04    2024-11-12          Paid  
1997              0.00 2024-11-04    2024-11-18          Paid  
1998              0.00 2024-11-04    2024-11-28          Paid  
1999              0.00 2024-11-04    2024-11-08          Paid  

[2000 rows x 9 columns]

<span style="color: pink"> For handling missing values in the Summenebenkosten (additional costs) field, we adopted a conservative approach by replacing null values with zero. While the ideal practice would involve consultation with the BI and sales teams to understand why these values were missing, we determined this to be a reasonable assumption given the relatively small number of affected records (only 2 missing values) and the fact that additional costs often represent optional or incidental charges. This approach preserves all records while minimizing potential impact on aggregate financial reporting.</span>

<span style="color: pink"> Regarding the ZahlungsbetragBrutto (gross payment amount) calculations, we implemented a derived calculation using the formula SummeNetto × (1 + MwStSatz/100) for cases where values were either null or zero. This mathematical approach ensures consistency with known VAT rates while maintaining the original values where they existed. The calculation follows standard accounting principles where gross amounts can be reliably determined from net amounts and applicable tax rates. We maintained an audit trail of these calculated values for transparency. </span>

<span style="color: pink"> For the payment date (Zahlungsdatum) field containing 403 missing values, we introduced a binary PaymentStatus classification (Paid/Unpaid) as an interim solution. In a production environment, we would absolutely need to collaborate with finance teams to understand whether missing dates truly represent unpaid invoices or potentially reflect data entry delays, system issues, or alternative payment arrangements. The current implementation serves as a pragmatic placeholder for analytical purposes, with clear documentation of this assumption to prevent misinterpretation of the payment status classification.</span>

In [41]:
invoice_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 9 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   ReNummer              2000 non-null   int64         
 1   SummeNetto            2000 non-null   float64       
 2   MwStSatz              2000 non-null   int64         
 3   ZahlungsbetragBrutto  2000 non-null   float64       
 4   KdNr                  2000 non-null   int64         
 5   Summenebenkosten      2000 non-null   float64       
 6   ReDatum               2000 non-null   datetime64[ns]
 7   Zahlungsdatum         1597 non-null   datetime64[ns]
 8   PaymentStatus         2000 non-null   object        
dtypes: datetime64[ns](2), float64(3), int64(3), object(1)
memory usage: 140.8+ KB


#### <span style="color:orange">**Customer Dataset**</span>

In [42]:
customer_dataset

id   Kdnr                                      Verlagsname     Region
0        5  20172                                   1. FC Nürnberg   Nürnberg
1       19  20137  Allgäuer Zeitung / Allgäuer Zeitungsverlag GmbH   Bodensee
2       27  20115                            Augsburger Allgemeine    München
3       69  10113                         Brainpool TV Productions  Rheinland
4       72  10154                 RFW / Redaktionsbüro Wipperfürth    Hamburg
..     ...    ...                                              ...        ...
596  14014  81391                                              PPF        NaN
597  14021  81398          VLAAMSE RADIO- EN TELEVISIEOMROEP (VRT)        NaN
598  14025  81402                           Umweltinstitut München        NaN
599  14031  81408                       Everprod - Groupe Elephant        NaN
600  14034  81411                                  Lennart Homeyer        NaN

[601 rows x 4 columns]

In [46]:
customer_dataset['Region'] = customer_dataset['Region'].apply(lambda x: 'Not Provided' if pd.isna(x) else x)
customer_dataset

id   Kdnr                                      Verlagsname  \
0        5  20172                                   1. FC Nürnberg   
1       19  20137  Allgäuer Zeitung / Allgäuer Zeitungsverlag GmbH   
2       27  20115                            Augsburger Allgemeine   
3       69  10113                         Brainpool TV Productions   
4       72  10154                 RFW / Redaktionsbüro Wipperfürth   
..     ...    ...                                              ...   
596  14014  81391                                              PPF   
597  14021  81398          VLAAMSE RADIO- EN TELEVISIEOMROEP (VRT)   
598  14025  81402                           Umweltinstitut München   
599  14031  81408                       Everprod - Groupe Elephant   
600  14034  81411                                  Lennart Homeyer   

           Region  
0        Nürnberg  
1        Bodensee  
2         München  
3       Rheinland  
4         Hamburg  
..            ...  
596  Not Provided  
597  Not Provided  
598  Not Provided  
599  Not Provided  
600  Not Provided  

[601 rows x 4 columns]

In [47]:
customer_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 601 entries, 0 to 600
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   id           601 non-null    int64 
 1   Kdnr         601 non-null    int64 
 2   Verlagsname  601 non-null    object
 3   Region       601 non-null    object
dtypes: int64(2), object(2)
memory usage: 18.9+ KB


#### <span style="color:orange">**Position Dataset**</span>

In [48]:
position_dataset

id    ReId     KdNr  Nettobetrag   Bildnummer   VerDatum
0       4154495  102038  50292.0        19.00  168161043.0 2022-09-29
1       4154496  102038  50292.0        19.00  108514408.0 2022-09-12
2       4154497  102038  50292.0        19.00   55374589.0 2022-09-08
3       4154498  102038  50292.0        19.00   55374589.0 2022-09-08
4       4154499  102038  50292.0        19.00   57865918.0 2022-09-08
...         ...     ...      ...          ...          ...        ...
129087  5726949  103209  10813.0        12.00  100000000.0 2025-01-27
129088  5726950  103209  10813.0        12.00  100000000.0 2025-02-25
129089  5726951  103209  10813.0        12.00  100000000.0 2025-02-25
129090  5726952  103209  10813.0        12.00  100000000.0 2025-02-23
129091  5726953  103209  10813.0         0.24  100000000.0 2025-02-15

[129092 rows x 6 columns]

In [53]:
position_null = position_dataset.isnull().sum()
position_null

id             0
ReId           0
KdNr           1
Nettobetrag    1
Bildnummer     1
VerDatum       4
dtype: int64

In [55]:
missing_rows_position = position_dataset[position_dataset['Nettobetrag'].isna() | 
                                         position_dataset['Bildnummer'].isna() | 
                                         position_dataset['VerDatum'].isna()]
missing_rows_position

id    ReId     KdNr  Nettobetrag   Bildnummer VerDatum
482     4156318  102055  79640.0        800.0  100000000.0      NaT
84350   4250726  103259  11257.0         75.0  165070078.0      NaT
84413   4250725  103254  20258.0         80.0  169144734.0      NaT
124096  5715403  103381      NaN          NaN          NaN      NaT

In [56]:
position_dataset['VerDatum'] = position_dataset['VerDatum'].apply(lambda x: 'Not Provided' if pd.isna(x) else x)
position_dataset

position_dataset = position_dataset.dropna(subset=['KdNr'])

In [57]:
position_dataset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 129091 entries, 0 to 129091
Data columns (total 6 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   id           129091 non-null  int64  
 1   ReId         129091 non-null  int64  
 2   KdNr         129091 non-null  float64
 3   Nettobetrag  129091 non-null  float64
 4   Bildnummer   129091 non-null  float64
 5   VerDatum     129091 non-null  object 
dtypes: float64(3), int64(2), object(1)
memory usage: 6.9+ MB


<span style="color: pink">To maintain data quality while preserving records where possible, we implemented two key cleaning steps for the positions dataset. First, for missing usage dates (VerDatum), we replaced null values with the label "Not Provided" rather than dropping records, as the date field, while useful for analysis, is not critical for core reporting. This ensures all position records remain in the dataset while clearly flagging incomplete entries. </span>

<span style="color: pink"> Second, we removed rows where the customer ID (KdNr) was missing, as this field is essential for attributing revenue and linking to customer records. Given that only a small number of rows (1 in this case) lacked a customer ID, deleting them had minimal impact on the dataset's overall integrity. This approach prioritizes referential accuracy ensuring all positions can be tied to valid customers, while transparently handling missing supplementary data like dates. Both steps align with standard data governance practices, balancing completeness with reliability.</span>

### <span style="color:#ADD8E6">**Tasks**</span>

#### <span style="color:orange">**How many positions are linked to invoices that are missing payment info**</span>

In [59]:
invoice_payment_mered = invoice_dataset.merge(position_dataset, left_on='ReNummer', right_on='ReId', how='left')
invoice_payment_mered

unpaid_positions = invoice_payment_mered[invoice_payment_mered['PaymentStatus'] == 'Unpaid']

ReNummer  SummeNetto  MwStSatz  ZahlungsbetragBrutto  KdNr_x  \
0         103601     11786.7         0              11786.70   78911   
1         103601     11786.7         0              11786.70   78911   
2         103601     11786.7         0              11786.70   78911   
3         103601     11786.7         0              11786.70   78911   
4         103601     11786.7         0              11786.70   78911   
...          ...         ...       ...                   ...     ...   
126784    101617       172.0         7                184.04   20361   
126785    101617       172.0         7                184.04   20361   
126786    101617       172.0         7                184.04   20361   
126787    101617       172.0         7                184.04   20361   
126788    101617       172.0         7                184.04   20361   

        Summenebenkosten    ReDatum Zahlungsdatum PaymentStatus         id  \
0                    0.0 2025-04-04           NaT        Unpaid  4516297.0   
1                    0.0 2025-04-04           NaT        Unpaid  4516298.0   
2                    0.0 2025-04-04           NaT        Unpaid  4516299.0   
3                    0.0 2025-04-04           NaT        Unpaid  4516300.0   
4                    0.0 2025-04-04           NaT        Unpaid  4516301.0   
...                  ...        ...           ...           ...        ...   
126784               0.0 2024-11-05           NaT        Unpaid  4256809.0   
126785               0.0 2024-11-05           NaT        Unpaid  5616785.0   
126786               0.0 2024-11-05           NaT        Unpaid  5616786.0   
126787               0.0 2024-11-05           NaT        Unpaid  5616787.0   
126788               0.0 2024-11-05           NaT        Unpaid  5616788.0   

            ReId   KdNr_y  Nettobetrag    Bildnummer             VerDatum  
0       103601.0  78153.0       741.64  1.000000e+08  2023-01-15 00:00:00  
1       103601.0  78152.0        14.68  1.130208e+08  2023-01-15 00:00:00  
2       103601.0  78152.0        14.68  8.709228e+07  2023-01-15 00:00:00  
3       103601.0  78152.0        14.68  1.306617e+08  2023-01-15 00:00:00  
4       103601.0  78152.0        14.68  8.709228e+07  2023-01-15 00:00:00  
...          ...      ...          ...           ...                  ...  
126784  101617.0  79290.0        -3.60  1.000000e+08  2022-09-15 00:00:00  
126785  101617.0  20361.0        43.00  1.047484e+09  2024-08-15 00:00:00  
126786  101617.0  20361.0        43.00  1.047484e+09  2024-08-15 00:00:00  
126787  101617.0  20361.0        43.00  1.362015e+07  2024-08-15 00:00:00  
126788  101617.0  20361.0        43.00  5.145465e+06  2024-08-15 00:00:00  

[18104 rows x 15 columns]

In [60]:
total_unpaid_positions = len(unpaid_positions) 
total_unpaid_invoices = unpaid_positions['ReNummer'].nunique()  

print(f"Total positions linked to unpaid invoices: {total_unpaid_positions}")
print(f"Distinct invoices with unpaid positions: {total_unpaid_invoices}")

Total positions linked to unpaid invoices: 18104
Distinct invoices with unpaid positions: 403


<span style="color: pink"> Our identification of positions linked to unpaid invoices specifically targets records with missing payment dates (Zahlungsdatum), which serves as the most reliable indicator of non-payment. This approach captures invoices that are either (a) genuinely unpaid by customers or (b) lack payment confirmation due to systemic data gaps. </span>


<span style="color: pink">  Notably, we excluded three computationally repaired records from this count (1 missing ZahlungsbetragBrutto and 2 missing Summenebenkosten values), as these were automatically derived using business rules rather than representing confirmed unpaid amounts. While this provides a conservative baseline for collections tracking, the true unpaid total could be marginally higher (by approximately 0.5%) if including these imputed values. For comprehensive reporting, we recommend supplementing these results with a separate "data-repaired amounts" category and collaborating with Finance to validate whether invoices with calculated values require manual review. This dual approach ensures analytical rigor while maintaining transparency about data quality limitations. </span>

#### <span style="color:orange">**How much revenue is attributed to placeholder media ID '100000000'**</span>

In [61]:
placeholder_positions = position_dataset[position_dataset['Bildnummer'] == 100000000]
placeholder_revenue = placeholder_positions['Nettobetrag'].sum()


print(f"Revenue from placeholder media: €{placeholder_revenue:,.2f}")

Revenue from placeholder media: €1,319,897.91


#### <span style="color:orange">**How many invoices have no positions attached**</span>

In [64]:
invoices_present = invoice_dataset[~invoice_dataset['ReNummer'].isin(position_dataset['ReId'])] 

ReNummer  SummeNetto  MwStSatz  ZahlungsbetragBrutto   KdNr  \
1283    102318       316.0         7                338.12  20743   
1626    101975       200.0         0                200.00  79630   

      Summenebenkosten    ReDatum Zahlungsdatum PaymentStatus  
1283               0.0 2025-01-03    2025-01-08          Paid  
1626               0.0 2024-12-03           NaT        Unpaid

In [65]:
print(f"Total invoices without positions: {len(invoices_present)}")

Total invoices without positions: 2


Add some explanations here